src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/GetCoveragesValue.java

```diff
@@ -0,0 +1,230 @@
+            }
+            reader = ci.getGridCoverageReader(null, null);
+            reader2d = (GridCoverage2DReader) reader;
+            if (reader2d.getOriginalEnvelope().contains(position)) {
+                try {
+                    gridCoverage2D = getGridCoverage2D(reader2d, position);
+                } catch (TransformException e) {
+                    throw new WPSException("Could not transform point " + position, e);
+                }
+                evaluationValue = convertPrimitiveArrayToNumberArray(gridCoverage2D.evaluate(position));
+                nodata = checkNoData(gridCoverage2D, evaluationValue);
+                if (!nodata) {
+                    break;
+                }
+            }
+        }
+        if (gridCoverage2D == null || nodata) {
+            return new ValuesAtPoint(new Double[] {});
+        }
+
+        return new ValuesAtPoint(evaluationValue);
+    }
+
+    private static GridCoverage2D getGridCoverage2D(GridCoverage2DReader reader, Position position)
+            throws IOException, TransformException {
+        GridCoverage2D gridCoverage2D;
+
+        Format format = reader.getFormat();
+        final ParameterValueGroup readParams = format.getReadParameters();
+        final Map<String, Serializable> parameters = CoverageUtils.getParametersKVP(readParams);
+
+        // Calculate the bbox around the point
+        MathTransform worldToGridCorner =
+                reader.getOriginalGridToWorld(PixelInCell.CELL_CORNER).inverse();
+        Coordinate coordinate = new Coordinate(position.getCoordinate()[0], position.getCoordinate()[1]);
+        JTS.transform(coordinate, coordinate, worldToGridCorner);
+        GridEnvelope2D testRange =
+                new GridEnvelope2D((int) (coordinate.getX() - 1), (int) (coordinate.getY() - 1), 2, 2);
+
+        // Calculate the world bounds of the buffered point
+        final MathTransform gridToWorldCorner = reader.getOriginalGridToWorld(PixelInCell.CELL_CORNER);
+        final GeneralBounds worldBounds = CRS.transform(gridToWorldCorner, new GeneralBounds(testRange.getBounds()));
+        worldBounds.setCoordinateReferenceSystem(reader.getCoordinateReferenceSystem());
+
+        // make sure mosaics with many superimposed tiles won't blow up with
+        // a "too many open files" exception
+        String maxAllowedTiles = ImageMosaicFormat.MAX_ALLOWED_TILES.getName().toString();
+        if (parameters.keySet().contains(maxAllowedTiles)) {
+            parameters.put(maxAllowedTiles, 1);
+        }
+
+        // Since the read sample image won't be greater than 5x5 pixels and we are limiting the
+        // number of granules to 1, we may do direct read instead of using JAI
+        String useJaiImageRead = ImageMosaicFormat.USE_JAI_IMAGEREAD.getName().toString();
+        if (parameters.keySet().contains(useJaiImageRead)) {
+            parameters.put(useJaiImageRead, false);
+        }
+        parameters.put(
```

Comment:
Again, please try to keep related code related together. If you're computing the grid geometry at the beginning of the method, finish the job of setting up the grid geometry parameter there.

Reply:
done